Feat: Add minimal support for i18n

dearchap opened the pull request:
<!--
  This template provides some ideas of things to include in your PR description.
  To start, try providing a short summary of your changes in the Title above.
  If a section of the PR template does not apply to this PR, then delete that section.
 -->

## What type of PR is this?

_(REQUIRED)_

<!--
  Delete any of the following that do not apply:
 -->

- feature

## What this PR does / why we need it:

_(REQUIRED)_

<!--
  What goal is this change working towards?
  Provide a bullet pointed summary of how each file was changed.
  Briefly explain any decisions you made with respect to the changes.
  Include anything here that you didn't include in *Release Notes*
  above, such as changes to CI or changes to internal methods.
-->

Enable basic internalization support for urfave/cli

## Which issue(s) this PR fixes:

_(REQUIRED)_

<!--
If this PR fixes one of more issues, list them here.
One line each, like so:

Fixes #123
Fixes #39
-->

Fixes #980 

## Special notes for your reviewer:

_(fill-in or delete this section)_

<!--
   Is there any particular feedback you would / wouldn't like?
   Which parts of the code should reviewers focus on?
-->

## Testing

_(fill-in or delete this section)_

<!--
  Describe how you tested this change.
-->

## Release Notes

_(REQUIRED)_
<!--
  If this PR makes user facing changes, please describe them here. This
  description will be copied into the release notes/changelog, whenever the
  next version is released. Keep this section short, and focus on high level
  changes.

  Put your text between the block. To omit notes, use NONE within the block.
-->

```release-note

This adds basic framework support for internationalization for urfave/cli. Right now the languages built in are en-US/en-GB which zero changes between the messages. As people request more languages the capability to add this is present with no changes to core code base
```

meatballhat commented:
Ack sorry about the conflicts! 😭

meatballhat reviewed (commented):
- This is very exciting! 🎉 
- It looks like the generated catalog code sets the `message.DefaultCatalog` package-level variable 🤔 Does this mean that this library is intended for use at the _application_ level and not from a dependency like `github.com/urfave/cli/v3`?
- Is there supposed to be an import of `_ "github.com/urfave/cli/v3/internal/translations"` somewhere in the top level so that the catalog is initialized?
- I had to manually install `gotext` to get this working locally. Can a `gotext`-ensuring command be added to `internal/build/build.go` like the one that ensures `goimports` exists?

dearchap commented:
@meatballhat

- Addressed all the conflicts
- Added import of _ "github.com/urfave/cli/v3/internal/translations"
- Added ensure-gotext to internal build 
- Added ensure-gotext and go-generation to the build system

meatballhat commented:
@dearchap Thank you! I sincerely appreciate your work on this 🤘🏼 

I'm still confused about the intended usage given [this line in the generated code](
https://github.com/dearchap/cli/blob/d43b64899ad6275591450859aac28458995d87bf/internal/translations/catalog.go#L38). How is this supposed to work if `urfave/cli/v3` is being used by an application that _also_ uses code generated by `gotext`?

Given that [the template file for the generated code](https://github.com/golang/text/blob/e50348080f29449bcd6808c11400b3d45f08b09d/message/pipeline/generate.go#L321) hasn't been touched in ~4 years, I'm wondering if using `golang.org/x/text/message/pipeline` (via `gotext`) is the right move 🤔

dearchap commented:
@meatballhat you have a keen eye. I tested cli with these changes along with a sample app that uses its own catalog and basically the translation doesn't work as one catalog overrides the other due to the message.DefaultCatalog in init. I'm going to redo this PR.

dearchap commented:
@meatballhat I've hacked around the package global Catalog initialization. I dont like it but this is what we have until support for multiple Catalogs is rolled into the text/message package itself. 
https://cs.opensource.google/go/x/text/+/refs/tags/v0.10.0:message/catalog/catalog.go;l=281

For now this is the simplest route without adding dependencies on other i18n libraries which look to me more heavyweight

dearchap commented:
@meatballhat Also I'm not sure why only the window tests are failing. Can you take a look ?

meatballhat commented:
> @meatballhat Also I'm not sure why only the window tests are failing. Can you take a look ?

Offhand I don't see why that failure is happening 🤔 Can the `make go-generate` step be limited to most recent Go on ubuntu like with some of the nearby steps?

meatballhat commented on internal/translations/dcatalog.go:
Is the intent here that the load order of the files is by name and that using `init` functions to stash the `message.DefaultCatalog` as `oldCatalog` will allow us to keep a reference to the catalog that is "ours"?

dearchap replied:
Yes correct.

meatballhat replied:
OK. hmm. Is this going to work reliably? Is load order and how it relates to `init` func running even guaranteed? (my quick search isn't encouraging 😅)

For me, this all tells a story of `golang.org/x/text/message` *not really being ready* for production usage, and yet there's no specific warning against using it in production. I'm confused!

dearchap replied:
https://medium.com/@markbates/go-init-order-dafa89fcef22

dearchap replied:
@meatballhat i cannot guarantee the load order. So I think we should just scrap this approach.
